File: udp_client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

#define PORT 29527
#define SERVER_IP "10.23.149.5"
#define BUFFER_SIZE 1024

int main() {
    int sock = 0;
    struct sockaddr_in serv_addr;
    char *json_message = "{\"name\": \"Fengpangzi\"}";
    char buffer[BUFFER_SIZE] = {0};

    // Create a TCP socket
    if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
        printf("Socket creation error\n");
        return -1;
    }

    // Server address setup
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(PORT);
    serv_addr.sin_addr.s_addr = inet_addr(SERVER_IP);

    // Send JSON message to the server
    if (sendto(sock, json_message, strlen(json_message), 0, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0) {
        printf("Send message failed\n");
        close(sock); // Close the client socket
        return -1;
    }
    printf("Message sent to server : %s\n", json_message);

    // Optionally, read the server's response
//    int valread = read(sock, buffer, BUFFER_SIZE);
//    printf("Server response: %s\n", buffer);

    close(sock); // Close the client socket
    return 0;
}

File: udp_client.py
import socket
import json

# Server details
SERVER_IP = '10.23.149.5'  # Replace with the actual IP address of the server
SERVER_PORT = 29527           # The port the server is listening on

# JSON data to send
data = {
    "name": "FreeRTOS Client",
    "message": "Hello from Python!"
}

# Convert the dictionary to a JSON string
json_data = json.dumps(data)

def udp_client():
    # Create a TCP/IP socket
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    
    try:
        # Connect the socket to the server
        client_socket.connect((SERVER_IP, SERVER_PORT))
        print(f"Connected to server at {SERVER_IP}:{SERVER_PORT}")

        # Send the JSON data
        client_socket.sendto(json_data.encode('utf-8'), (SERVER_IP, SERVER_PORT))
        print(f"Sent JSON data: {json_data}")

        # Optionally, receive a response from the server
#        response = client_socket.recv(1024).decode('utf-8')
#        print(f"Received from server: {response}")

    except Exception as e:
        print(f"Error: {e}")

    finally:
        # Close the socket
        client_socket.close()
        print("Connection closed.")

if __name__ == '__main__':
    udp_client()
